### ISSUE TITLE
[Feature] Enum types from strings or metadata

### ISSUE THREAD

**Issue by tick-rick** [fsprojects/SwaggerProvider#146]
[Here is a schema.json](https://github.com/fsprojects/SwaggerProvider/files/4507635/schema.txt) that demonstrates the issue.

When "enum" schema components are encountered in json schema, SwaggerProvider seems to strip their info away and just label the type as strings or int (depending on JsonStringEnumConverter). 

This makes working with APIs a bit painful. Azure seems to deal with this issue using [x-ms-enum](https://github.com/stankovski/AutoRest/blob/master/Documentation/swagger-extensions.md#x-ms-enum) which means it would add meta data for codegen (regardless of if its an int or string encoding for enum value).

Code to demonstrate issue, noting that it is not possible to even create a TaskType, one has to work out what the valid values are by looking at the raw JSON schema (and hoping it doesn't change like it did for me):
```
open SwaggerProvider

type Api = OpenApiClientProvider<"schema.json">
let client = Api.Client()
client.GetApiAdminTest("aosdijoasjid") |> ignore
```
It would be nice if SwaggerProvider could generate this:
`type TaskType = This | That | Other` based on x-ms-enum or string 

### Related information 

NET Core SDK & Linux

**Comment by tick-rick** [comment]
@sergey-tihon I made a first pass at this [here](https://github.com/tick-rick/SwaggerProvider/commit/119018741f065849eef02e4a739efe0cc5bb0291#diff-1d120bdef869b2cd319a1a4b9358b048) but I am having trouble in a few areas and need pointers (I only took a guess at how to do it):

- I needed to remove net461 to get compile working on Linux, so I dont think I can PR yet 
- Where do I need to add in the code that handles serialization? e.g. Enum -> String?
- Testing feels very light and I feel I am not adding them in the right place (I left it commented out on my fork for now)

**Comment by sergey-tihon** [comment]
> It would be nice if SwaggerProvider could generate this:
type TaskType = This | That | Other based on x-ms-enum or string

First of all, I am not surer that it will be easy  task to generate provided enums
https://github.com/fsprojects/FSharp.TypeProviders.SDK/issues/264 AFAIK, [SqlEnumProvider.fs](https://github.com/fsprojects/FSharp.Data.SqlClient/blob/master/src/SqlClient.DesignTime/SqlEnumProvider.fs) is best starting point but I cannot guarantee that it will work as you expect.

> Where do I need to add in the code that handles serialization? e.g. Enum -> String?

Serialization is done here https://github.com/fsprojects/SwaggerProvider/blob/master/src/SwaggerProvider.Runtime/ProvidedApiClientBase.fs#L34-L37 but you even can inherit from ProvidedClient and experiment with serialization without modifying TP code http://fsprojects.github.io/SwaggerProvider/#/Customization#serialization

> Testing feels very light and I feel I am not adding them in the right place (I left it commented out on my fork for now)

Is it possible to add new controller here https://github.com/fsprojects/SwaggerProvider/tree/master/tests/Swashbuckle.WebApi.Server/Controllers and integrate `AutoRest Extensions for Swagger 2.0` with Swashbuckle to emit `x-ms-enum`?

**Comment by tick-rick** [comment]
@sergey-tihon I am very close to getting it to work. There are two failings tests, both with the same root cause (very similar to [this issue](https://github.com/fsprojects/SwaggerProvider/issues/140)), where Some(X) is serialized when types are non primitive. I will try track this down. 

```
15:47:04 DBG] All/v3/Swashbuckle.UpdateControllers.Tests.Update Enum GET Test starting... <Expecto>
      Request starting HTTP/1.1 GET http://localhost:5000/api/UpdateEnum?x=Some(Absolute)  
```

- I updated tests to correctly test enum values (previously test code used int).
- Added support in for `values` field of x-ms-enum
- [added a Filter](https://github.com/tick-rick/SwaggerProvider/blob/master/tests/Swashbuckle.WebApi.Server/Filters/EnumExtensionFilter.fs) to Swashbuckle.WebApi.Server/ to generate x-ms-enum.
- Latest code still lives on https://github.com/tick-rick/SwaggerProvider as I cant test with net46 targetframework


x-ms-enum in swagger.json:
![Screenshot_2020-04-30 Screenshot](https://user-images.githubusercontent.com/50384876/80662512-c28de380-8ad4-11ea-844b-911ca2a360a5.png)

**Comment by abelbraaksma** [comment]
Just got hit by this myself, in the sense that I thought my OpenApi was wrong as I expected a `DU` or something like that to be generated from the `enum` values in the OpenApi yaml specification.

I appreciate that this isn't trivial to add (just mapping it to .NET enums clearly won't work, though DU's seem a natural fit). Just good to know this isn't supported atm.

**Comment by petrkoutnycz** [comment]
Just leaving a comment here that handling enums in a strongly-typed fashion would be nice.
For instance, I have a string enum defined in Swagger and it is basically skipped.
<img width="388" alt="image" src="https://user-images.githubusercontent.com/9723402/207565378-dcbc069f-a355-4d7e-9632-ea9f774c525f.png">

**Comment by abelbraaksma** [comment]
Just realising this, but isn’t generating DUs from a TP currently simply not supported? I mean, CLI enums seems to the best we can do here, right?

**Comment by sergey-tihon** [comment]
> Just realising this, but isn’t generating DUs from a TP currently simply not supported?

 Yes, all F# specific types are not supported (like DUs and records), that is why it is `int` for now.

Not sure about CLI enums ... Never saw them emitted by TP.

**Comment by petrkoutnycz** [comment]
CLI enums would be better than nothing, that is for sure. In my case (see the picture above), I can assign just pure string - not DU and not even CLI enum.
